Adding support for VpcConnectorEgressSettings field to cloudfunctions…

jambt opened the pull request:
…. Also Updating the website docs for terraform.

* Adding support for VpcConnectorEgressSettings field to cloudfunctions.

* Updating the webiste docs for terraform.

```release-note:enhancement
`cloudfunctions`: added support for `vpc_connector_egress_settings` to `google_cloudfunctions_function`
```

jambt commented:
Hi there,

I followed the PR #5981 to implement a missing setting for the vpc egress. Feel free to criticize or improve, since this is my first contribution.

rileykarson commented on google/resource_cloudfunctions_function.go:
```diff
@@ -33,8 +33,16 @@ var allowedIngressSettings = []string{
+var allowedVpcConnectorEgressSettings = []string{
+	"ALL_TRAFFIC",
+	"PRIVATE_RANGES_ONLY",
+	"VPC_CONNECTOR_EGRESS_SETTINGS_UNSPECIFIED",
```
```suggestion
```

GCP API docs are slightly misleading, the "UNSPECIFIED" option in enums will never actually appear and will generally be rejected by the API server.

rileykarson commented on google/resource_cloudfunctions_function.go:
```diff
@@ -180,6 +188,13 @@ func resourceCloudFunctionsFunction() *schema.Resource {
+			"vpc_connector_egress_settings": {
+				Type:         schema.TypeString,
+				Optional:     true,
+				Default:      functionDefaultVpcConnectorEgressSettings,
```
```suggestion
```

See above- this defaults to unset in some cases, and `PRIVATE_RANGES_ONLY` in others.

rileykarson commented on website/docs/r/cloudfunctions_function.html.markdown:
```diff
@@ -133,6 +133,8 @@ Eg. `"nodejs8"`, `"nodejs10"`, `"python37"`, `"go111"`.
 
 * `vpc_connector` - (Optional) The VPC Network Connector that this cloud function can connect to. It can be either the fully-qualified URI, or the short name of the network connector resource. The format of this field is `projects/*/locations/*/connectors/*`.
 
+* `vpc_connector_egress_settings` - (Optional) String value that controls what traffic is send through the VPC Connector. Allowed values are ALL_TRAFFIC and PRIVATE_RANGES_ONLY
```
```suggestion
* `vpc_connector_egress_settings` - (Optional) The egress settings for the connector, controlling what traffic is diverted through it. Allowed values are `ALL_TRAFFIC` and `PRIVATE_RANGES_ONLY`. Defaults to `PRIVATE_RANGES_ONLY`. If unset, this field preserves the previously set value.
```

rileykarson commented on website/docs/d/datasource_cloudfunctions_function.html.markdown:
```diff
@@ -54,6 +54,7 @@ exported:
 * `ingress_settings` - Controls what traffic can reach the function.
 * `labels` - A map of labels applied to this function.
 * `service_account_email` - The service account email to be assumed by the cloud function.
+* `vpc_connector_egress_settings` - Controls which traffic is send through the VPC connector
```
```suggestion
* `vpc_connector_egress_settings` - The egress settings for the connector, controlling what traffic is diverted through it.
```

rileykarson commented on google/resource_cloudfunctions_function.go:
```diff
@@ -180,6 +188,13 @@ func resourceCloudFunctionsFunction() *schema.Resource {
 
+			"vpc_connector_egress_settings": {
+				Type:         schema.TypeString,
+				Optional:     true,
```
```suggestion
				Optional:     true,
				Computed:     true,
```

This field is conditionally set to `PRIVATE_RANGES_ONLY`, which Terraform can't express well. Marking as `Computed` or "ignore if unset" is most appropriate.

jambt commented:
Hi @rileykarson,

thank you for the very valid corrections. I implemented them all. Furthermore, I removed the line `const functionDefaultVpcConnectorEgressSettings = "VPC_CONNECTOR_EGRESS_SETTINGS_UNSPECIFIED"` since it is now obsolete.

Can something further be improved?
